pkg/webhook/resources/virtualmachine/validator.go

```diff
@@ -69,12 +75,134 @@ func (v *vmValidator) Resource() types.Resource {
 	}
 }
 
+func (v *vmValidator) getClusterNetworkForNad(nwName string) (clusterNetwork string, err error) {
+	words := strings.Split(nwName, "/")
+	var namespace, name string
+	switch len(words) {
+	case 1:
+		namespace, name = "default", words[0]
+	case 2:
+		namespace, name = words[0], words[1]
+	default:
+		return clusterNetwork, fmt.Errorf("invalid network name %s", nwName)
+	}
+
+	nad, err := v.nadCache.Get(namespace, name)
+	if err != nil {
+		return clusterNetwork, err
+	}
+	clusterNetwork, ok := nad.Labels[keyClusterNetwork]
+	if !ok {
+		return clusterNetwork, err
+	}
+
+	return clusterNetwork, nil
+}
+
+func (v *vmValidator) getVMInterfaceInfo(vms []*kubevirtv1.VirtualMachine) (MacAddrPerCluster map[string]map[string]string, err error) {
+	MacAddrPerCluster = make(map[string]map[string]string) //map[clusterName]map[macaddr]intfName
+
+	for _, vm := range vms {
+		vmName := vm.Namespace + "/" + vm.Name
+
+		if vm.Spec.Template == nil {
+			return MacAddrPerCluster, fmt.Errorf("vm %s template is nil", vm.Name)
+		}
+
+		vmCluster := make(map[string]string) // map[netIntfName]clusterNetwork
+		vmNetworks := vm.Spec.Template.Spec.Networks
+		for _, vmNetwork := range vmNetworks {
+			if vmNetwork.Multus == nil || vmNetwork.Multus.NetworkName == "" {
+				continue
+			}
+			cn, err := v.getClusterNetworkForNad(vmNetwork.Multus.NetworkName)
+			if err != nil {
+				return MacAddrPerCluster, err
+			}
+			vmCluster[vmNetwork.Name] = cn
+		}
+
+		vmInterfaces := vm.Spec.Template.Spec.Domain.Devices.Interfaces
+
+		var macAddrs map[string]string
+		for _, vmInterface := range vmInterfaces {
+			clusNet := vmCluster[vmInterface.Name]
+			if _, exists := MacAddrPerCluster[clusNet]; !exists {
+				macAddrs = make(map[string]string)
+			} else {
+				macAddrs = MacAddrPerCluster[clusNet]
+			}
+			macAddrs[vmInterface.MacAddress] = vmInterface.Name + "+" + vmName
+			MacAddrPerCluster[clusNet] = macAddrs
+		}
+	}
+
+	return MacAddrPerCluster, nil
+}
+
+func (v *vmValidator) checkForDuplicateMacAddrs(vm *kubevirtv1.VirtualMachine) (err error) {
+	var newvm []*kubevirtv1.VirtualMachine
+	var oldvmsInfo map[string]map[string]string
+	var oldvmsList []*kubevirtv1.VirtualMachine
+
+	//get map[clustername]map[mac]string info for new vm
+	newvm = append(newvm, vm)
+	newVMInfo, err := v.getVMInterfaceInfo(newvm)
+	if err != nil {
+		return err
+	}
+
+	if vm.Spec.Template == nil {
+		return fmt.Errorf("vm %s template is nil", vm.Name)
+	}
+
+	vmInterfaces := vm.Spec.Template.Spec.Domain.Devices.Interfaces
+
+	for _, vmInterface := range vmInterfaces {
+		oldvms, err := v.vmCache.GetByIndex(indexerwebhook.VMByMacAddress, vmInterface.MacAddress)
```

Comment:
another question: the index VMByMacAddress is composed of MAC list, and use a single `vmInterface.MacAddress` to get, will it return?

Reply:
vmByMacaddrs groups the vms based on mac address with mac address as key/index.
When user tries to give key as any mac address, the output returns list of vms matching that key if it matches or empty vm list.